docs/blueprints/referral-manager-v1.md

```diff
@@ -0,0 +1,196 @@
+## Rationale
+
+We could add a new column `referral_tokens` in `users` table to store user's unredeemed tokens, since the `registration_token` is a temporary table.
+
+## Implementation
+
+- Create a private repository for Referral Manager.
+- Create `tokens` and `satellites` table in Referral Manager database.
+- Create a `Delete` method for satellite `registration_tokendb`.
+- Implementing an endpoint on satellite for gathering users whose current count of remaining invitation token is 0.
+- Implementing generating invitation tokens from Referral Manager CLI.
+- Implementing an endpoint on satellite for saving new referral links into users table.
+- Implementing an endpoint on Referral Manager for verifying invitation tokens.
+- Replace existing registration token logic.
+
+Pseudocode
+
+The Token struct represents both a Go struct (on satellite and Referral Manager) as well as the schema for the `tokens` table on the Referral Manager
+
+```
+type Token struct {
+	Secret [32]byte
+	OwnerID uuid.UUID
+	RedeemedBy uuid.UUID
+	HostSatelliteURL string
+    RedeemedSatelliteURL string
+	Status string
+}
+```
+Statuses: `unsent` (host satellite doesn't know about it yet), `unredeemed` (host satellite knows about it but it has not been used), `redeemed` (someone has used this referral link to register alreadys)
+
+Referral Manager endpoints:
+```
+// GetEligibleUsers retrieves a list of users who have 0 remaining invitation tokens
+// from each satellite provided, and saves those users in memory
+GetEligibleUsers([]satelliteURLs) map[satelliteURL]int {
+    eligibleUsers := make(map[satelliteURL][]UserID)
+    eligibleUsersCounts := make(map[satelliteURL]int)
+    for _, satellite := range satelliteURL {
+       users := satellite.GetUsersReferral() 
+        eligibleUsers[satellite] = users
+        eligibleUsersCounts[satellite]++
+    }
+    referralManager.eligibleUsers = eligibleUsers
+    return eligibleUsersCounts
+}
+
+// SendInvitationTokens generates tokens, saves those in the referral manager db
+// and then sends newly created tokens to each respective satellite
+GenerateAndSendInvitationTokens([]satelliteURL, tokensPerUser int) map[satelliteURL]int {
+    eligibleUsers := referralManager.eligibleUsers
+    if eligibleUsers == nil {
+        return Error.New("No users to generate tokens for. Make sure to run GetEligibleUsers first")
+    }
+    var tokens []Token
+    for satellite, users := range eligibleUsers {
+        for _, user := range users {
+            for i:=0; i<tokensPerUser; i++ {
+                newToken := Token{data: generateRandomToken(), user: user, satellite: satellite}
+                db.CreateToken(newToken)
+                tokens = append(tokens, newToken)
+            }
+        }
+    }
+
+    results := make(map[satelliteURL]int)
+    for _, satellite := range satelliteURLs {
+        successCount := satellite.AddInvitationTokens(tokens)
+        results[satellite] = successCount
+    }
+    return results
+}
```

Comment:
I noticed you are sharing state across calls in `referralManager.eligibleUsers`.  Is this necessary? I'm assuming you are doing this so the list of users they agreed to is what gets processed. But, what if 2 people are running these referrals at the same time?  Is it safe enough to just generate the list again in GenerateAndSendInvitationTokens since it's not likely it changed much since GetEligibleUsers was called (unless the another user ran referrals at the same time)? This would also reduce the amount of re-generations.

Maybe `GetEligibleUsers` could return a count and hash of the IDs to be processed and the CLI could pass the hash to `GenerateAndSendInvitationTokens` so that when the list gets generated again, it can be hashed and compared.
If they don't match, `GenerateAndSendInvitationTokens` could return an error.

Or if it is not critical for the user list to be exact, the CLI could just push the user count they agreed to back to `GenerateAndSendInvitationTokens`.

Reply:
Good point! In this case, I don't think it's critical for the user list to be exact. As long as we can make sure that the total user count is the same, we can proceed with the `GenerateAndSendInvitationTokens`. I will update the code to check the user count instead. 
What do you mean by `This would also reduce the amount of re-generations.` ? From what I understand, the generated user list will be saved in memory. If another user ran `GetEligibleUsers` at the same time, as long as we have the same total user count, it should be fine. Is it not a good idea to run it again?

Reply:
Discussed this with @VinozzZ and we think we could solve this problem the following way:
`GetEligibleUsers` checks `referralManger.eligibleUsers` before starting. If `referralManager.eligibleUsers` is not nil, it means one of two things:
1. Someone else is using the referral manager at the same time and has not generated/sent tokens yet (very unlikely)
2. The CLI was ctrl+Cd between the get users step and the generate/send tokens step (more likely)

So if `GetEligibleUsers` is called and `referralManager.eligibleUsers` is set to something other than nil, we return an error and the above information is conveyed to the CLI user. Then the CLI user has the option to abort or to continue anyway. If they decide to continue, we call a new endpoint like `ClearUsers` which sets `eligibleUsers` to nil and start over  with `GetEligibleUsers`.

We just need to make sure `GenerateAndSendInvitationTokens sets `referralManager.eligibleUsers` to nil when it finishes.

Reply:
Works for me since this is really an internal tool anyway.

Reply:
i just updated the doc to reflect this